File: swift/Sources/DSA/Algorithms/QuickSort.swift
/// Quick sort algorithm with median-of-three pivot selection.
/// Time: O(n log n) average, O(n²) worst case
/// Space: O(log n) average for recursion stack
/// Stable: No
public enum QuickSort {
    /// Sort an array in-place.
    public static func sort<T: Comparable>(_ array: inout [T], by comparator: ((T, T) -> Bool)? = nil) {
        let compare = comparator ?? { $0 < $1 }
        quickSort(&array, 0, array.count - 1, by: compare)
    }

    /// Return a sorted copy of the array.
    public static func sorted<T: Comparable>(_ array: [T], by comparator: ((T, T) -> Bool)? = nil) -> [T] {
        var result = array
        sort(&result, by: comparator)
        return result
    }

    private static func quickSort<T>(_ array: inout [T], _ low: Int, _ high: Int, by compare: (T, T) -> Bool) {
        guard low < high else { return }

        let pivotIndex = partition(&array, low, high, by: compare)
        quickSort(&array, low, pivotIndex - 1, by: compare)
        quickSort(&array, pivotIndex + 1, high, by: compare)
    }

    private static func partition<T>(_ array: inout [T], _ low: Int, _ high: Int, by compare: (T, T) -> Bool) -> Int {
        // Median-of-three pivot selection
        let mid = low + (high - low) / 2

        if compare(array[mid], array[low]) { array.swapAt(low, mid) }
        if compare(array[high], array[low]) { array.swapAt(low, high) }
        if compare(array[mid], array[high]) { array.swapAt(mid, high) }

        let pivot = array[high]
        var i = low - 1

        for j in low..<high {
            if compare(array[j], pivot) || (!compare(pivot, array[j])) {
                i += 1
                array.swapAt(i, j)
            }
        }

        array.swapAt(i + 1, high)
        return i + 1
    }
}

File: kotlin/src/main/kotlin/algorithms/QuickSort.kt
package algorithms

/**
 * Quick sort algorithm with median-of-three pivot selection.
 * Time: O(n log n) average, O(n^2) worst case
 * Space: O(log n) average for recursion stack
 * Stable: No
 */
object QuickSort {
    /**
     * Sort a mutable list in-place.
     */
    fun <T : Comparable<T>> sort(list: MutableList<T>, comparator: Comparator<T>? = null) {
        quickSort(list, 0, list.lastIndex, comparator)
    }

    /**
     * Return a sorted copy of the list.
     */
    fun <T : Comparable<T>> sorted(list: List<T>, comparator: Comparator<T>? = null): List<T> {
        val result = list.toMutableList()
        sort(result, comparator)
        return result
    }

    private fun <T : Comparable<T>> quickSort(
        list: MutableList<T>,
        low: Int,
        high: Int,
        comparator: Comparator<T>?
    ) {
        if (low < high) {
            val pivotIndex = partition(list, low, high, comparator)
            quickSort(list, low, pivotIndex - 1, comparator)
            quickSort(list, pivotIndex + 1, high, comparator)
        }
    }

    private fun <T : Comparable<T>> partition(
        list: MutableList<T>,
        low: Int,
        high: Int,
        comparator: Comparator<T>?
    ): Int {
        // Median-of-three pivot selection
        val mid = low + (high - low) / 2

        if (compare(list[mid], list[low], comparator) < 0) swap(list, low, mid)
        if (compare(list[high], list[low], comparator) < 0) swap(list, low, high)
        if (compare(list[mid], list[high], comparator) < 0) swap(list, mid, high)

        val pivot = list[high]
        var i = low - 1

        for (j in low until high) {
            if (compare(list[j], pivot, comparator) <= 0) {
                i++
                swap(list, i, j)
            }
        }

        swap(list, i + 1, high)
        return i + 1
    }

    private fun <T> swap(list: MutableList<T>, i: Int, j: Int) {
        val temp = list[i]
        list[i] = list[j]
        list[j] = temp
    }

    private fun <T : Comparable<T>> compare(a: T, b: T, comparator: Comparator<T>?): Int =
        comparator?.compare(a, b) ?: a.compareTo(b)
}
